torch/_custom_op.py

```diff
@@ -370,3 +449,128 @@ def validate_function_matches_schema(
             f"Schema has kwonlyarg names {kwonlyarg_names} but function has "
             f"{arg_spec.kwonlyargs}."
         )
+
+
+# Global dictionary holding weak references to all CustomOp objects
+# Used to query the CustomOp associated with a specific C++ dispatcher operator.
+# An example usage is FakeTensor: FakeTensor checks if a specific operator
+# has an implementation registered via the CustomOp API.
+global_registry: weakref.WeakValueDictionary = weakref.WeakValueDictionary({})
+
+
+def get_none():
+    return None
+
+
+global_ctx_getter: typing.Callable = get_none
+
+
+# NOTE [ctx inside the fake implementation]
+# If a user has an operator with data-dependent output shape, then when writing
+# a fake implementation they must query the current ctx and use methods on the
+# ctx to construct a new unbacked symint.
+#
+# This is done via us setting the global_ctx_getter function every time a fake
+# implementation is invoked.
+def get_ctx() -> "FakeTensorImplCtx":
+    """get_ctx() returns the current FakeTensorImplCtx object.
+
+    Calling ``get_ctx()`` is only valid inside of a fake implementation.
+    """
+    return global_ctx_getter()
+
+
+@contextlib.contextmanager
+def set_ctx_getter(ctx_getter):
+    global global_ctx_getter
+    prev = global_ctx_getter
+    try:
+        global_ctx_getter = ctx_getter
+        yield
+    finally:
+        global_ctx_getter = prev
+
+
+class FakeTensorImplCtx:
+    """
+    Context object for writing FakeTensor rules for custom operators.
+    """
+
+    def __init__(self, _shape_env, _op):
+        self._shape_env = _shape_env
+        self._op = _op
+
+    def new_data_dependent_symint(self, *, min=2, max=None) -> torch.SymInt:
+        """Constructs a new symint (symbolic int) representing a data-dependent value.
+
+        This is useful for writing the fake implementation (which is necessary
+        for torch.compile) for a CustomOp where an output Tensor has a size
+        that depends on the data of the input Tensors.
+
+        Args:
+            min (int): A statically known inclusive lower bound for this symint.
+                min must be at least 2 due to implementation details of
+                torch.compile. Default: 2.
+            max (Optional[int]): A statically known inclusive upper bound for this
+                symint. Default: None
+
+        .. warning:
+
+            It is important that the ``min`` and ``max`` (if not None) values are set
+            correctly, otherwise, there will be undefined behavior under
+            torch.compile. The default value of ``min`` is 2.
```

Comment:
Is it worth mentioning that the dfeault of 2 is due to 0/1 specialization? (is that right?)

Reply:
Yes. It's worth mentioning, I'll update it